shadowsocks/udp.go

```diff
@@ -195,15 +174,46 @@ func (s *udpService) Serve(clientConn net.PacketConn) error {
 				}
 				debugUDPAddr(clientAddr, "Got location \"%s\"", clientLocation)
 
+				ip := clientAddr.(*net.UDPAddr).IP
+				var cipher shadowaead.Cipher
+				unpackStart := time.Now()
+				textData, keyID, cipher, err = unpack(ip, textBuf, cipherData, s.ciphers)
+				timeToCipher = time.Now().Sub(unpackStart)
+
+				if err != nil {
+					return onet.NewConnectionError("ERR_CIPHER", "Failed to unpack initial packet", err)
+				}
+
 				udpConn, err := net.ListenPacket("udp", "")
 				if err != nil {
 					return onet.NewConnectionError("ERR_CREATE_SOCKET", "Failed to create UDP socket", err)
 				}
 				targetConn = nm.Add(clientAddr, clientConn, cipher, udpConn, clientLocation, keyID)
+			} else {
+				unpackStart := time.Now()
```

Comment:
This is duplicated. Can we move it to outside the if-then-else?

Reply:
We can, but that will also include the GeoIP lookup time, socket creation, etc.

Alternatively, we could define a helper function like
```go
func time(f func()) time.Duration {
    before := time.Now()
    f()
    return time.Now().Sub(before)
}
```
That's more verbose, but it makes the parallelism explicit.

Reply:
Oh, nevermind, they are different unpack calls.